src/lib/signature_lib/checked.ml

```diff
@@ -235,13 +268,35 @@ module Schnorr
     in
     go (Curve.Scalar.length_in_bits - 1) Curve.zero
 
-  let verify ((s, h) : Signature.t) (pk : Public_key.t) (m : Message.t) =
-    let pre_r = shamir_sum (s, Curve.one) (h, pk) in
-    if Curve.equal Curve.zero pre_r then false
-    else
-      let r = compress pre_r in
-      let h' = Message.hash ~nonce:r m in
-      Curve.Scalar.equal h' h
+  let sign (d : Private_key.t) m =
+    let k_prime =
+      ((Curve.Scalar.to_bits d) @ (Message.to_bits m))
+          |> Blake2.bits_to_string
+          |> Random_oracle.digest_string
+          (* TODO : should Random_oracle.digest_bits exist ?
+           * currently only Random_oracle.Checked.digest_bits is defined *)
+          |> Curve.Scalar.of_bits
+    in
+    (* TODO : should this be an if/else instead ? *)
+    assert (not (Curve.Scalar.(equal k_prime zero))) ;
+    let r_point = Curve.scale Curve.one k_prime in
+    let yr = get_y r_point in
+    let k =
+    if is_even yr then k_prime
+    else Curve.Scalar.(order - k_prime) in
+    let e = (get_x r_point) @ (compress (Curve.scale Curve.one d)) @ (Message.to_bits m) in
+    let s = Curve.Scalar.(k + (of_bits e) * d) in
+    (get_x r_point, s)
+
+  let verify ((r, s) : Signature.t) (pk : Public_key.t) (m : Message.t) =
+    assert (Curve.is_on_curve pk) ;
```

Comment:
No need to check this here, this will (or ought to be :) ) be checked of any public key as it's deserialized. Also assertions like this could fail on a bad input, and we don't want the program to crash in that case, we probably want this function to return `false.`